hs_composite_resource/models.py

```diff
@@ -577,6 +578,67 @@ def get_missing_file_type_metadata_info(self):
                                               'missing_elements': missing_elements})
         return metadata_missing_info
 
+    def get_data_services_urls(self):
+        """
+        Generates data services URLs for the resource.
+
+        If the resource contains any GeoFeature or GeoRaster content, and if it's public,
+        generate data service endpoints.
+        """
+
+        if self.raccess.public is True:
+            resource_data_types = [lf.data_type for lf in self.logical_files]
+
+            if any(
+                data_type in [
+                    'GeographicFeature', 'GeographicRaster'
+                ] for data_type in resource_data_types
+            ):
+                wms_url = (
+                    f'{settings.HS_GEOSERVER}/wms?'
+                    f'service=WMS&'
+                    f'version=1.3.0&'
+                    f'request=GetCapabilities&'
+                    f'namespace=HS-{self.short_id}'
+                )
+            else:
+                wms_url = None
+
+            if 'GeographicFeature' in resource_data_types:
+                wfs_url = (
+                    f'{settings.HS_GEOSERVER}/wfs?'
+                    f'service=WFS&'
+                    f'version=1.1.0&'
+                    f'request=GetCapabilities&'
+                    f'namespace=HS-{self.short_id}'
+                )
+            else:
+                wfs_url = None
+
+            if 'GeographicRaster' in resource_data_types:
+                wcs_url = (
+                    f'{settings.HS_GEOSERVER}/wcs?'
+                    f'service=WCS&'
+                    f'version=1.1.0&'
+                    f'request=GetCapabilities&'
+                    f'namespace=HS-{self.short_id}'
+                )
+            else:
+                wcs_url = None
+
+        else:
+            wms_url = None
+            wfs_url = None
+            wcs_url = None
+
+        data_services_urls = {
+            'wms_url': wms_url,
+            'wfs_url': wfs_url,
+            'wcs_url': wcs_url
+        }
+
```

Comment:
What happens when these services are down/broken? If appropriate, ensure a friendly error is passed to the enduser when clicking on Preview and something logged to help ops staff troubleshoot the hydroshare/hydroshare systems

Reply:
f214a44: Added some basic error handling and logging. We don't really have a way right now for HydroShare to verify that each link is working as expected, but this should catch issues with how GeoServer is configured on HydroShare's end.

Reply:
If I understand, the URL will always appear in the UI, and the worst case scenario is the user clicks on it and gets a 404/500 or other page load error. Perhaps RENCI could add these URLs to what gets monitored in production. Not sure what systems we use to monitor such things. Just a thought. I believe the contacts for that would be Chris C. and Dan.

Reply:
If the URL would sometimes appear as an Error, then some message should be displayed in its place to email help@